design-proposals/klip-3-serialization-of-single-fields.md

```diff
@@ -0,0 +1,1076 @@
+ ```
+
+* A new ``serialization.rst`` will be added to the developer guide:
+
+```rst
+_ksql_serialization:
+
+KSQL Serialization
+==================
+
+.. _ksql_formats
+
+=======
+Formats
+=======
+
+KSQL currently supports three serialization formats:
+
+*. ``DELIMITED`` supports comma separated values. See :ref:`delimited_format` below.
+*. ``JSON`` supports JSON values. See :ref:`json_format` below.
+*. ``AVRO`` supports AVRO serialized values. See :ref:`avro_format` below.
+
+
+.. _delimited_format
+
+---------
+DELIMITED
+---------
+
+The ``DELIMITED`` format supports comma separated values.
+
+The serialized object should be a Kafka-serialized string, which will be split into columns.
+
+For example, given a KSQL statement such as:
+
+.. code:: sql
+
+    CREATE STREAM x (ID BIGINT, NAME STRING, AGE INT) WITH (VALUE_FORMAT='DELIMITED', ...);
+
+KSQL splits a value of ``120, bob, 49`` into the three fields with ``ID`` of ``120``,
+``NAME`` of ``bob`` and ``AGE`` of ``49``.
+
+This data format supports all KSQL :ref:`data types <data-types>` except for container types, 
+i.e. ``ARRAY``, ``MAP`` and ``STRUCT``.
+
+.. _json_format
+
+----
+JSON
+----
+
+The ``JSON`` format supports JSON values.
+
+The JSON format supports all of KSQL's ref:`data types <data-types>`. As JSON does not itself
+support a map type, KSQL serializes ``MAP``s as JSON objects.  Because of this, the JSON format
+can only support ``MAP`` objects that have ``STRING`` keys.
+
+The serialized object should be a Kafka-serialized string containing a valid JSON value. The format
+supports JSON objects and top-level primitives, arrays and maps. See below for more info.
+
+JSON Objects
+------------
+
+Values that are JSON objects are probably the most common.
+
+For example, given a KSQL statement such as:
+
+.. code:: sql
+
+    CREATE STREAM x (ID BIGINT, NAME STRING, AGE INT) WITH (VALUE_FORMAT='JSON', ...);
+
+And a JSON value of:
+
+.. code:: json
+
+       {
+         "id": 120,
+         "name": "bob",
+         "age": "49"
+       }
+
+KSQL will deserialize the JSON object's fields into the corresponding fields of the stream.
+
+.. note:: Where the key or value schema contains only a single field KSQL, by default, 
+          will both expect source data and produce sink data where the field has been 
+          serialized as an anonymous value, rather than as a named field within an 
+          JSON object.
+          
+          For more information, see :ref:`single_field_wrapping`.
+
+Top-level Primitives
+--------------------
+
+The JSON format supports reading and writing top-level JSON primitives, but only if 
+the target schema contains a single field of a compatible type.
+
+For example, given a KSQL statement with only a single field in the value schema:
+
+.. code:: sql
+
+    CREATE STREAM x (ID BIGINT) WITH (VALUE_FORMAT='JSON', ...);
```

Comment:
We should probably add `WRAP_SINGLE_VALUES=false` into this example, since this won't work out of the box (unless the user happened to set `ksql.persistence.wrap.single.values=false` in their config).

Same comment for "Top-level Arrays" and "Top-level Maps" below, and for "Top-level primitives, arrays and maps" in the Avro section.

Reply:
updated.